### PR TITLE
Apply CSSRule insertion/deltetion on Flush event

### PR DESCRIPTION
**Pull request by VladimirMilenko**
Currently, during the node replacement, the `sheet` property is being automatically assigned to null. 
In order to support virtual sheet mutation, we need to store that sheet in another structure.
We also cannot reuse the existing reference to sheet, as the value at this reference would be removed at the `replaceChild` call.

To avoid that, we will create:
```ts
  private virtualStyleRulesMap!: Map<INode, CSSRule[]>;
```
This rule array map would keep the latest CSSRule array for each node during the virtual processing of events and would ensure, that at the Flush event, `style` node has exact same array of css rules, as the one in memory.

I've also added a test which is able to check, whether the stylesheet rule exists in the sheet after insertion.

Fixes https://github.com/rrweb-io/rrweb/issues/525

### CODE REVIEW

**Comment by VladimirMilenko** [comment]
cc @Juice10
This change is removing the check you added. Could you provide some info, whether that might create an issue?

**Comment by Yuyz0112** [comment]
LGTM

@Mark-Fenng PTAL

**Comment by Mark-Fenng** [comment]
A great way to fix the issue.
BTW, is placeholderNode still necessary to exist?

File: test/replayer.test.ts [hunk 1]
```diff
@@ -146,6 +146,19 @@ describe('replayer', function (this: ISuite) {
...
+
+    const result = await this.page.evaluate(`
+      const rules = Array.from(replayer.iframe.contentDocument.head.children)
+        .filter(x=>x.nodeName === 'STYLE')
+        .reduce((acc, node) => {
+          acc.push(...node.sheet.rules);
+          return acc;
+        }, []);
+
+        rules.some(x=>x.selectorText === ".css-1fbxx79")
```
Comment (Juice10) [inline test/replayer.test.ts:158]: Would be nice if we could simplify this a little
```suggestion
      const rules = [...replayer.iframe.contentDocument.styleSheets].map((sheet)=> sheet.rules);
      rules.some(x=>x.selectorText === ".css-1fbxx79")
```

File: test/replayer.test.ts [hunk 2]
```diff
@@ -170,6 +183,31 @@ describe('replayer', function (this: ISuite) {
...
```
Comment (Juice10) [inline test/replayer.test.ts:205]: Same goes for here
```suggestion
      const rules = [...replayer.iframe.contentDocument.styleSheets].map((sheet)=> sheet.rules);
      rules.some(x=>x.selectorText === ".css-1fbxx79")
```

**Review by Juice10 — CHANGES_REQUESTED** [review]
Really loving this, great work @VladimirMilenko

**Comment by Juice10** [comment]
> A great way to fix the issue.
> BTW, is placeholderNode still necessary to exist?

As far as I can remember it is necessary the first time the `.sheet` is accessed.
But for performance reasons it would be advisable skip the placeholder/adding-stylesheet-to-dom part when the node already exists in the `virtualStyleRulesMap`.

**Comment by Juice10** [comment]
It looks like constructible stylesheets (`new CSSStyleSheet`) is still in the [proposal phase](https://wicg.github.io/construct-stylesheets/) and is only implemented in Chrome, and [Firefox (behind experimental flag)](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Experimental_features#constructable_stylesheets).
In Safari I couldn't get it working directly but I was able to access it via a workaround: `Object.create(CSSStyleSheet.prototype)`

We might be able to get it working by using this polyfill: https://github.com/calebdwilliams/construct-style-sheets

**Comment by Juice10** [comment]
@VladimirMilenko feel free to close this PR. I ended up incorporating most of it in https://github.com/rrweb-io/rrweb/pull/618 which fixes this issue.

**Pull request closed without merging**